posts/share-cpp/index.qmd

```diff
@@ -0,0 +1,244 @@
+
+1. Include the autogenerated file `mypackage_RcppExports.h`; this is already done as part of the package header generation.
+2. Directly write C++ code in the package header. This is technically possible, but unlikely to be a good option as your package's C++ codebase grows.
+3. Manually include any other C++ header files in the package header. This last option might lead to a package header such as that shown below.
+
+```{.cpp filename="include/inst/mypackage.h"}
+// Manually created package header with manual code inclusion
+#ifndef mypackage_mypackage_H
+#define mypackage_mypackage_H
+
+// include files using paths relative to inst/include
+#include "header_01.h"
+#include "header_02.h"
+
+#endif  // mypackage_mypackage_H
+```
+
+Here, the header files might contain code that you wish to make available to other packages, such as a C++ function, struct, or class, and indeed in the current package as well --- more on how to do this below.
+
+## Using Rcpp in header code
+
+Using {Rcpp}'s C++ functionality, such as the Rcpp classes `DataFrame` or `List`, or classes and functions of Rcpp-based packages such as {RcppEigen}, is as simple as including those headers in the appropriate location, just as one would in a source file --- see the example below.
+
+```{.cpp filename="inst/include/header_01.h"}
+// In a manually created header file, say, header_01.h
+// which is included in mypackage.h
+
+// to use Rcpp
+#include <Rcpp.h>
+
+// note the use of inline, more on this later
+inline void hello_world_rcpp() {
+  Rcpp::Rcout << "hello world!\n";
+}
+```
+
+The appropriate headers are automatically included in autogenerated package headers' helper files, and the developer need not do anything more.
+
+::: {.callout-tip}
+Don't forget to link to {Rcpp} or similar packages to the package under development by adding the package names under `Imports`, `Depends`, or `LinkingTo` as appropriate.
+
+This can often be handled by functions in the {usethis} package such as `usethis::use_rcpp_eigen()`.
+You might also need to add `// [[Rcpp::depends(<package>)]]` in your package's C++ source files, with a suitable package dependency specified.
+:::
+
+The same principles apply to using C++ code from this package ({mypackage}) in future packages.
+
+## Using header code in the package
+
+There are some considerations when seeking to use header code from {mypackage} within {mypackage} itself.
+
+Any functions defined in the package headers must be inline functions (see the example above).
+This prevents compilation errors related to multiple definitions.
+
+C++ source files should include the package header, using `#include mypackage.h`.
+Functions, structs, or classes defined in header files will be available from the namespace `mypackage`, as shown in the example below.
+
+The code in header files will usually need to be wrapped in (R)C++ code that is exposed to R to make functions from the headers available in R --- see the snippet below.
+
+```c++
```

Comment:
Could you add a filename hint here as well please?

Reply:
Added in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/34/commits/8dd883c7a4a763d2f58e2446bd764bced6488033